src/main/java/com/amazon/aws/iot/greengrass/semver/SemVer.java

```diff
@@ -0,0 +1,340 @@
+        if (!matcher.find()) {
+            throw new IllegalArgumentException("Version \"" + version + "\" is not valid");
+        }
+        MatchResult m = matcher.toMatchResult();
+
+        this.raw = version;
+
+        // these are actually numbers
+        this.major = Integer.parseInt(m.group(1));
+        this.minor = Integer.parseInt(m.group(2));
+        this.patch = Integer.parseInt(m.group(3));
+
+        if (this.major < 0) {
+            throw new IllegalArgumentException("Invalid major version");
+        }
+
+        if (this.minor < 0) {
+            throw new IllegalArgumentException("Invalid minor version");
+        }
+
+        if (this.patch < 0) {
+            throw new IllegalArgumentException("Invalid patch version");
+        }
+
+        // numberify any prerelease numeric ids
+        if (m.group(4) != null) {
+            this.prerelease = Arrays.stream(m.group(4).split("\\.", -1)).map((String id) -> {
+                if (IS_NUMERIC.matcher(id).find()) {
+                    int num = Integer.parseInt(id);
+                    if (num >= 0) {
+                        return num;
+                    }
+                }
+                return id;
+            }).collect(Collectors.toList());
+        } else {
+            this.prerelease = new ArrayList<>();
+        }
+
+        this.build = m.group(5) == null ? new ArrayList<>() : Arrays.asList(m.group(5).split("\\.", -1));
+        this.version = this.format();
+    }
+
+    protected String format() {
+        String v = String.format("%s.%s.%s", this.major, this.minor, this.patch);
+        if (this.prerelease != null && this.prerelease.size() > 0) {
+            v += String.format("-%s",
+                    this.prerelease.stream().map(Objects::toString).collect(Collectors.joining(".")));
+        }
+        this.version = v;
+        return v;
+    }
+
+    /**
+     * Get the normalized value of the SemVer including prerelease and build
+     * @return String.
+     */
+    public String getValue() {
+        if (build.isEmpty()) {
+            return version;
+        }
+        return String.format("%s+%s", version, String.join(".", build));
+    }
+
+    @Override
+    public int compareTo(SemVer other) {
+        if (other.version.equals(this.version)) {
+            return 0;
+        }
+
+        int mainCompared = this.compareMain(other);
+        if (mainCompared != 0) {
+            return mainCompared;
+        }
+        return this.comparePre(other);
+    }
+
+    public int compareMain(SemVer other) {
+        int maj = compareIdentifiers(this.major, other.major);
+        if (maj != 0) {
+            return maj;
+        }
+        int min = compareIdentifiers(this.minor, other.minor);
+        if (min != 0) {
+            return min;
+        }
+        return compareIdentifiers(this.patch, other.patch);
+    }
+
+    public int comparePre(SemVer other) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 13. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 149-155 into a separate method.